### ISSUE TITLE
MNLI BERT Transformer Model Is Not Using Token Type IDs

### ISSUE THREAD

**Issue by nelson-liu** [allenai/allennlp#5129]
## Checklist

- [x] I have verified that the issue exists against the `main` branch of AllenNLP.
- [x] I have read the relevant section in the [contribution guide](https://github.com/allenai/allennlp/blob/main/CONTRIBUTING.md#bug-fixes-and-new-features) on reporting bugs.
- [x] I have checked the [issues list](https://github.com/allenai/allennlp/issues) for similar or identical bug reports.
- [x] I have checked the [pull requests list](https://github.com/allenai/allennlp/pulls) for existing proposed fixes.
- [x] I have checked the [CHANGELOG](https://github.com/allenai/allennlp/blob/main/CHANGELOG.md) and the [commit log](https://github.com/allenai/allennlp/commits/main) to find out if the bug was already fixed in the main branch.
- [x] I have included in the "Description" section below a traceback from any exceptions related to this bug.
- [x] I have included in the "Related issues or possible duplicates" section beloew all related issues and possible duplicate issues (If there are none, check this box anyway).
- [x] I have included in the "Environment" section below the name of the operating system and Python version that I was using when I discovered this bug.
- [x] I have included in the "Environment" section below the output of `pip freeze`.
- [x] I have included in the "Steps to reproduce" section below a minimally reproducible example.

## Description

I've been getting consistently low numbers with the MNLI BERT model ~1-2 accuracy points. It appears that the MNLI model is not using `token_type_ids`.

Here's what i see from allennlp

```
(Pdb) p tokens["tokens"]["type_ids"][0]
tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])
(Pdb) p tokens["tokens"]["mask"][0].long()
tensor([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])
(Pdb) p tokens["tokens"]["token_ids"][0].long()
tensor([  101,  1996,  2047,  2916,  2024,  3835,  2438,   102,  3071,  2428,
         7777,  1996, 14751,  6666,   102,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0])
```

Here's what I see with huggingface's run_glue script

```
(Pdb) input_ids[0]
tensor([  101,  1996,  2047,  2916,  2024,  3835,  2438,   102,  3071,  2428,
         7777,  1996, 14751,  6666,   102,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0])
(Pdb) attention_mask[0]
tensor([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0])
(Pdb) token_type_ids[0]
tensor([0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0])
```

The input IDs and attention mask match exactly (modulo padding), but the huggingface model provides token_type_ids while the AllenNLP model has them as all 0's.

This might also be an issue in question answering, I haven't checked.

## Environment

OS: Ubuntu 20.04

Python version: 
Python 3.7.10

## Steps to reproduce

1. Put a breakpoint at https://github.com/allenai/allennlp/blob/master/allennlp/modules/token_embedders/pretrained_transformer_embedder.py#L184 .
2. Run https://github.com/allenai/allennlp-models/blob/main/training_config/pair_classification/mnli_roberta.jsonnet , but with BERT instead of RoBERTa
3. Inspect the tokenizer output.

**Comment by nelson-liu** [comment]
Looks like the issue is here: https://github.com/allenai/allennlp/blob/main/allennlp/data/token_indexers/pretrained_transformer_indexer.py#L174-L176 . The token type IDs work just fine until they're overridden here.

When max_length is provided to the indexer (as it is in the config above), the token type IDs are overridden with all 0's. I'm not sure why this has to be the case. Seems like this was added this as part of the BART pr (https://github.com/allenai/allennlp/pull/4169/files#diff-1531e12f77c625639b9d3cab0a169259a0d92eaa43517738ea569e174d744513R181), maybe @Tobias-Rohde or @dirkgr could say more about what the rationale was (sorry, I realize this was quite awhile ago)?

**Comment by nelson-liu** [comment]
Indeed, if i remove the `max_length` from the indexer / embedder, things seem to work (note the 1's and 0's in the type_ids):

```
(Pdb) p tokens["tokens"]["type_ids"][0]
tensor([0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])
(Pdb) p tokens["tokens"]["mask"][0].long()
tensor([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])
(Pdb) p tokens["tokens"]["token_ids"][0]
tensor([  101,  1109,  1207,  2266,  1132,  3505,  1536,   102,  6064,  1541,
         7407,  1103, 15649,  6245,   102,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0])
```

**Comment by dirkgr** [comment]
Huh. That does not look like it happened by accident. But I can't imagine why?

**Comment by dirkgr** [comment]
I think I know now why. This is not an easy problem.

Setting `max_length` on the indexer makes it fold long sequences into multiple smaller sequences. For token ids, that's well defined: You start with 600 word pieces, so you get two sequences:
 1. `[CLS] + 510 content word pieces + [SEP]`
 2. `[CLS] + 90 content word pieces + [SEP]`

But for token type IDs, what should we expect? What's the correct token type id for a special token? The transformers are never trained on an example where the first content word piece already has a token type id of 1.

There is no correct answer for how to do token type ids with folding, so those two features are mutually exclusive.

Do we need such long sequences for NLI? Can we just turn off folding?

**Comment by epwalsh** [comment]
Why not keep the type IDs for regular tokens the same, and then just infer the type ID for each added special token based on the type ID of the adjacent regular token. For example:

```
Original non-folded sequence:

[CLS] a b c [SEP] d e f [SEP]
  0   0 0 0   0   1 1 1   1

Folds with a max length of 5 (3 regular tokens + 2 special tokens):

[CLS] a b c [SEP] | [CLS] d e f [SEP]
  0   0 0 0   0   |   1   1 1 1   1
```

This may not be the most technically correct way (like, BERT doesn't usually see a "[CLS]" token with type ID "1"), but at least we're not removing any information from the input.

**Comment by dirkgr** [comment]
I put a fix at https://github.com/allenai/allennlp/pull/5149.

That's not exactly what @epwalsh suggested. This instead pretends that the ids we got were part of a single sequence. That is almost always incorrect, but for some transformers it'll amount to the same thing, and either way it is better than losing the information completely.

**Comment by dirkgr** [comment]
Fixed in #5149.
